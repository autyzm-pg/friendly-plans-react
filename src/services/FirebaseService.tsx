import firebase from 'react-native-firebase';
import { Image } from 'react-native-image-crop-picker';

export class FirebaseService {
  /* Upload image to Firebase Storage and return public URL to download it. */
  static uploadUserImage = async (image: string): Promise<string> => {
    const user = firebase.auth().currentUser!;
    const imageRef = firebase
      .storage()
      .ref('users')
      .child(user.uid);
    const snapshot = await imageRef.putFile(image);
    return snapshot.downloadURL as string;
  };

  static updateUserImage = async (image: string): Promise<void> => {
    const photoURL = await FirebaseService.uploadUserImage(image);
    return firebase.auth().currentUser!.updateProfile({ photoURL });
  };
}
