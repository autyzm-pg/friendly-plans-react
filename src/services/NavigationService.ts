import { NavigationActions, NavigationContainerComponent, NavigationState } from 'react-navigation';

import { DialogProps } from 'models';

interface ScreenProps extends Partial<DialogProps> {
  [property: string]: any;
}

export default class NavigationService {
  navigator?: NavigationContainerComponent;

  setTopLevelNavigator = (navigatorRef: NavigationContainerComponent) => {
    this.navigator = navigatorRef;
  };

  navigate = (routeName: string, params?: ScreenProps) => {
    this.navigator!.dispatch(
      NavigationActions.navigate({
        routeName,
        params,
      }),
    );
  };

  goBack = () => {
    this.navigator!.dispatch(NavigationActions.back());
  };

  getActiveRouteName = (navigationState: NavigationState): string => {
    if (!navigationState) {
      return '';
    }
    const route = navigationState.routes[navigationState.index];
    // dive into nested navigators
    if (route.routes) {
      return this.getActiveRouteName(route as NavigationState);
    }
    return route.routeName;
  };
}
