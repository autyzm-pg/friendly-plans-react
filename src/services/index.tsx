import _AnalyticsService from './AnalyticsService';
import _CrashlyticsService from './CrashlyticsService';
import _NavigationService from './NavigationService';

export const NavigationService = new _NavigationService();
export const AnalyticsService = new _AnalyticsService(NavigationService);
export const CrashlyticsService = new _CrashlyticsService();
