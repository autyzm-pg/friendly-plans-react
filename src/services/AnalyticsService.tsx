import firebase from 'react-native-firebase';
import { NavigationState } from 'react-navigation';

import { User } from 'models';
import NavigationService from './NavigationService';

export default class AnalyticsService {
  navigationService: NavigationService;
  user: User | null;
  screenName?: string;

  constructor(navigationService: NavigationService) {
    firebase.analytics().setAnalyticsCollectionEnabled(!__DEV__);
    this.navigationService = navigationService;
    this.user = null;
  }

  setCurrentScreen = (prevState: NavigationState, currentState: NavigationState) => {
    const currentScreen = this.navigationService.getActiveRouteName(currentState);
    const prevScreen = this.navigationService.getActiveRouteName(prevState);

    if (prevScreen !== currentScreen) {
      firebase.analytics().setCurrentScreen(currentScreen, currentScreen);
      this.screenName = currentScreen;
    }
  };

  setCurrentUser = (user: User | null) => {
    this.user = user;
    firebase.analytics().setUserId(this.user ? this.user!.id.toString() : null);
  };
}
