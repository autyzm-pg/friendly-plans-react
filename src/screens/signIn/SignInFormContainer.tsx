import { Formik, FormikProps } from 'formik';
import React from 'react';
import firebase from 'react-native-firebase';
import * as Yup from 'yup';

import { i18n } from 'locale';
import { NavigationService } from 'services';
import { SignInForm } from './SignInForm';

export interface SignInFormData {
  email: string;
  password: string;
}

const initialValues = {
  email: '',
  password: '',
};

const validationSchema = Yup.object().shape({
  email: Yup.string()
    .email(i18n.t('validation:email'))
    .required(i18n.t('validation:required')),
  password: Yup.string()
    .min(6, i18n.t('validation:passwordLength'))
    .required(i18n.t('validation:required')),
});

interface State {
  loading: boolean;
}

export class SignInFormContainer extends React.PureComponent<{}, State> {
  state = {
    loading: false,
  };

  onSubmit = async (values: SignInFormData) => {
    this.setState({ loading: true });
    try {
      await firebase.auth().signInWithEmailAndPassword(values.email, values.password);
      NavigationService.navigate('Authenticated');
    } catch (error) {
      this.setState({ loading: false });
      NavigationService.navigate('Dialog', {
        title: i18n.t('common:error'),
        description: error.message,
      });
    }
  };

  navigateToResetPassword = () => {
    NavigationService.navigate('ResetPassword');
  };

  render() {
    return (
      <Formik
        initialValues={initialValues}
        onSubmit={this.onSubmit}
        validationSchema={validationSchema}
        render={this.renderForm}
      />
    );
  }

  renderForm = (props: FormikProps<SignInFormData>) => (
    <SignInForm {...props} loading={this.state.loading} navigateToResetPassword={this.navigateToResetPassword} />
  );
}
