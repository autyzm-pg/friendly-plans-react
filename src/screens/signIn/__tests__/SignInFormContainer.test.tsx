import { shallow } from 'enzyme';
import React from 'react';
import firebase from 'react-native-firebase';

import { NavigationService } from 'services';
import { SignInFormContainer } from '../SignInFormContainer';

describe('SignInFormContainer', () => {
  const props = {
    handleChange: jest.fn(),
    handleBlur: jest.fn(),
    values: {
      email: 'email@domain.com',
      password: 'topsecretpassword',
    },
    errors: {
      email: '',
      password: '',
    },
    touched: {
      email: false,
      password: false,
    },
    handleSubmit: jest.fn(),
  } as any;

  it('should match snapshot', () => {
    const wrapper = shallow(<SignInFormContainer />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render form correctly', () => {
    const wrapper = shallow(<SignInFormContainer />);
    const instance = wrapper.instance() as SignInFormContainer;
    const form = shallow(instance.renderForm(props));
    expect(form).toMatchSnapshot();
  });

  it('should navigate to authenticated on successful submit', async () => {
    const wrapper = shallow(<SignInFormContainer />);
    const instance = wrapper.instance() as SignInFormContainer;
    const data = {
      email: 'email@domain.com',
      password: 'topsecretpa$$w0rd',
    };
    const spy = jest.spyOn(NavigationService, 'navigate').mockImplementation(jest.fn());
    await instance.onSubmit(data);
    expect(spy).toBeCalledWith('Authenticated');
    jest.resetAllMocks();
  });

  it('should not navigate to authenticated on failed submit', async () => {
    const wrapper = shallow(<SignInFormContainer />);
    const instance = wrapper.instance() as SignInFormContainer;
    const data = {
      email: 'email@domain.com',
      password: 'topsecretpa$$w0rd',
    };
    // @ts-ignore
    firebase.auth = jest.fn().mockImplementationOnce(() => ({
      signInWithEmailAndPassword: jest.fn(() => Promise.reject('error')),
    }));
    const spy = jest.spyOn(NavigationService, 'navigate');
    await instance.onSubmit(data);
    expect(spy).not.toBeCalledWith('Authenticated');
    expect(spy).toBeCalledWith('Dialog', expect.objectContaining({}));
  });

  it('should navigate to reset password screen', () => {
    const wrapper = shallow(<SignInFormContainer />);
    const instance = wrapper.instance() as SignInFormContainer;
    const spy = jest.spyOn(NavigationService, 'navigate');
    instance.navigateToResetPassword();
    expect(spy).toBeCalledWith('ResetPassword');
  });
});
