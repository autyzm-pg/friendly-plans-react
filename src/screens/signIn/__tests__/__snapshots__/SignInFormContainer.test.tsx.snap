// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInFormContainer should match snapshot 1`] = `
<Formik
  enableReinitialize={false}
  initialValues={
    Object {
      "email": "",
      "password": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={
    ObjectSchema {
      "_blacklist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "_conditions": Array [],
      "_defaultDefault": [Function],
      "_deps": Array [],
      "_excludedEdges": Array [],
      "_exclusive": Object {},
      "_mutate": undefined,
      "_nodes": Array [
        "password",
        "email",
      ],
      "_options": Object {
        "abortEarly": true,
        "recursive": true,
      },
      "_type": "object",
      "_typeError": [Function],
      "_whitelist": RefSet {
        "list": Set {},
        "refs": Map {},
      },
      "fields": Object {
        "email": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
            "undefined": false,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
        "password": StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": true,
          },
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        },
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
    }
  }
/>
`;

exports[`SignInFormContainer should render form correctly 1`] = `
<View>
  <InputItem
    autoCapitalize="none"
    blurOnSubmit={true}
    error=""
    keyboardType="email-address"
    label="Email address"
    textContentType="emailAddress"
    touched={false}
    value="email@domain.com"
  />
  <InputItem
    blurOnSubmit={true}
    error=""
    label="Password"
    secureTextEntry={true}
    textContentType="password"
    touched={false}
    value="topsecretpassword"
  />
  <FlatButton
    buttonStyle={
      Object {
        "height": 36,
        "justifyContent": "flex-end",
      }
    }
    containerStyle={
      Object {
        "alignSelf": "flex-end",
        "top": -8,
      }
    }
    onPress={[Function]}
    title="forgot password?"
    titleStyle={
      Object {
        "fontFamily": "Rubik-Medium",
        "fontSize": 12,
        "letterSpacing": 0,
      }
    }
  />
  <Component
    backgroundColor="#6654EB"
    containerStyle={
      Object {
        "marginBottom": 20,
        "marginTop": 8,
      }
    }
    loading={false}
    onPress={[MockFunction]}
    title="Sign In"
  />
</View>
`;
